### ISSUE TITLE
Plan

### ISSUE THREAD

**Issue by turnage** [rgrpc/rgrpc#1]
# grpc

With a grpc implementation, rust services can drop-in replace non-rust services, increasing the amount of rust and decreasing the amount of non-rust in the world.

## Plot Zero -> Research

* Make sure we will retain rights to the work from our employers...
* Investigate what work is reusable and what to reinvent here.

## Plot One -> HTTP/2

grpc needs a (great) implementation of [HTTP/2](https://tools.ietf.org/html/rfc7540).

Prior art:

* https://github.com/mlalic/solicit

## Plot Two -> Protobuffer Compiler

grpc needs a (great) representation of protobuffers in rust.

Prior art:

* https://github.com/stepancheg/rust-protobuf

## Plot Three -> GRPC

[Website](http://www.grpc.io/).

[Required reading](http://platformlab.stanford.edu/Seminar%20Talks/gRPC.pdf).

Ongoing efforts:

* https://github.com/stepancheg/grpc-rust
* https://github.com/lambdastackio/grpc-rust

**Comment by mckeankylej** [comment]
In terms of https2 we have a couple of great starting points.
* The battle tested http library [hyper](https://github.com/hyperium/hyper)
* The up and coming tokio series of library based on futures-rs. Futures-rs is a game changer as it brings zero cost futures(promises) to rust. All your future code will compile down to a state machine as if you hand rolled the IO yourself. [tokio org](https://github.com/tokio-rs) [futures-rs](https://github.com/alexcrichton/futures-rs)

The proto buf story I suggest we wait for the chomp parsing library to stabilize as the other rust parsing library are good but once [chomp's] major prs get merged its going to be levels better than the competition. We might want to use [nom](https://github.com/Geal/nom) as its "byte-oriented" I like to think of it as "macro-oriented" as literately every parser is a macro. Its terrible to use but very very fast.

Overall I might just start developing the proto buf parser on chomps develop branch.

**Comment by mckeankylej** [comment]
+1 on 100% native. Its a great selling point. If we go that route we could write 100% safe code. As calling c library is inherently unsafe.

**Comment by mckeankylej** [comment]
What is core to grpc?

**Comment by turnage** [comment]
@mckeankylej 

Hyper [doesn't implement http/2](https://github.com/hyperium/hyper/issues/304). We should talk to a maintainer though; maybe we can help.

I think we should investigate the c binding route a little further. I'll do a small experiment and see how much sense it would make to depend on it and replace it incrementally with rust.

We should definitely use the prior art I listed for protobuf representation unless we have a solid reason not to.

See here for the [grpc core](https://github.com/grpc/grpc).

**Comment by mckeankylej** [comment]
Yea hyper doesn't have http2 support, we would be seriously giving back to the community by implementing that functionally. All the hyper devs are focused on rewriting hyper to use futures.

**Comment by turnage** [comment]
My participation is postponed until December 3rd.

**Comment by Kane-Sendgrid** [comment]
what do you guys plan to start from?

**Comment by mckeankylej** [comment]
I don't think we have any semblance of a plan yet. That being said I don't really understand what that question means.
